libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,324 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused parameters from function
+ * definition.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/YulString.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <optional>
+#include <variant>
+
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+
+namespace
+{
+
+bool anyFalse(vector<bool> const& _mask)
+{
+	return any_of(_mask.begin(), _mask.end(), [](bool b){ return !b; });
+}
+
+template<typename T>
+vector<T> applyBooleanMask(vector<T> const& _vec, vector<bool> const& _mask)
```

Comment:
From the name I would have thought it just computes an `&` and does not change the size of the vector. Maybe something like `cutOut`? But then the mask would have to be inverted...

Reply:
The name comes from Python datascience libraries: https://www.tensorflow.org/api_docs/python/tf/boolean_mask

Perhaps the name `booleanMask` can be confusing inside solidity. Would `filter` be a better choice?